storage/copy-to-s3: parquet format support with basic types

rjobanp opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

  * This PR adds a known-desirable feature. Fixes https://github.com/MaterializeInc/materialize/issues/26490

This only adds support for most scalar and simple byte/string types. I'll build on top of this to add support for more complex types going forward -- for Private Preview we will aim to support some subset of all possible MZ types (still TBD).

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]


    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

There is a large docstring and diagram in the 2nd commit on the `src/storage-operators/src/s3_oneshot_sink/parquet.rs` file that would be helpful to read at the beginning of your review.

The first commit implements support for creating a structure of Arrow Arrays from MZ rows. The second implements the parquet-format upload logic in the copy-to-s3 sink.

We chose to use the arrow-rs crate rather than arrow2/parquet2 (which is what persist currently uses) since after a conversation with @danhhz @bkirwi and @ParkMyCar it seemed non-trivial to couple the two implementations, and since arrow2 has been deprecated and merged into arrow-rs (https://github.com/jorgecarleitao/arrow2)

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,341 @@
+        let bucket = self.key_manager.bucket.clone();
+        let sdk_config = Arc::clone(&self.sdk_config);
+        let max_file_size = self.max_file_size;
+        let new_file = mz_ore::task::spawn(
+            || "ParquetFile::new",
+            ParquetFile::new(schema, bucket, object_key, sdk_config, max_file_size),
+        )
+        .wait_and_assert_finished()
```
`.run_in_task` from `FutureExt` might be easier!

rjobanp replied:
nice - will switch these all to that, much less code!

def- reviewed (commented):
Thanks for adding the tests. I'll add this to parallel-workload too and push a change later.

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,337 @@
+        if u64::cast_from(
+            self.writer.bytes_written()
+                + self.writer.in_progress_size()
+                + record_batch.get_array_memory_size(),
```
I think in general this specific check is fundamental to the correct operation of the sink, so each component should be documented! I think this is related to the above question I had about handling this error, which comes down: are we overcounting or undercounting, and how does that relate to the flow of the code itself?

guswynn reviewed (commented):
In general this is good, but I have lots of questions about the exact semantics of a couple things, particularly when we go over the file size limit

rjobanp commented on src/arrow-util/src/builder.rs:
```diff
@@ -0,0 +1,307 @@
+            let mut col_name = col_name.to_string();
+            // If we allow columns with the same name we encounter two issues:
+            // 1. The arrow crate will accidentally reuse the same buffers for the columns
+            // 2. Many parquet readers will error when trying to read the file metadata
+            // Instead we append a number to the end of the column name for any duplicates.
```
yep even a simple `COPY (SELECT 5::int 8::int) TO ... ` will result in two columns both named `int` in the `RelationDesc` we receive. 

I'll leave a TODO that we need to document this when we write the MZ docs

rjobanp commented on src/arrow-util/src/builder.rs:
```diff
@@ -0,0 +1,307 @@
+
+impl ArrowBuilder {
+    /// Initializes a new ArrowBuilder with the schema of the provided RelationDesc.
+    pub fn new(desc: &RelationDesc, default_capacity: usize) -> Result<Self, anyhow::Error> {
```
yep, adding a comment. It sets the default number of values (rows) that can be appended to each array builder before they need to reallocate

rjobanp commented on src/arrow-util/src/builder.rs:
```diff
@@ -0,0 +1,307 @@
+    pub fn to_record_batch(self) -> Result<RecordBatch, ArrowError> {
+        let mut arrays = vec![];
+        let mut fields = vec![];
+        for mut col in self.columns.into_iter() {
+            arrays.push(col.inner.finish());
+            fields.push(Field::new(col.field_name, col.data_type, col.nullable));
+        }
+        RecordBatch::try_new(Schema::new(fields).into(), arrays)
```
I'll adjust `to_record_batch` to use `fn schema` (`schema()` is used in the uploader to set the high-level schema of the writer, but is also needed as part of each record batch that is created)

rjobanp commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,337 @@
+/// Helper to tie the lifecycle of the `ArrowWriter` and `S3MultiPartUploader` together
+/// for a single parquet file.
+struct ParquetFile {
+    writer: ArrowWriter<Vec<u8>>,
+    uploader: S3MultiPartUploader,
```
That's definitely an interesting idea.. yes at the moment it would be difficult since `S3MultiPartUploader` is async, but the parquet crate does provide an async version `AsyncArrowWriter` https://docs.rs/parquet/latest/parquet/arrow/async_writer/struct.AsyncArrowWriter.html that requires something to implement `tokio::io::AsyncWrite`.

We could therefore implement `AsyncWrite` on `S3MultiPartUploader` to make this more elegant, though it seems somewhat non-trivial to do so (based on https://itnext.io/asyncwrite-and-a-tale-of-four-implementations-e63aef8397f7 ). I'm tempted to leave that as a TODO / potential future improvement if you agree

guswynn replied:
Yeah just leave it as a todo!

def- commented:
```
> COPY (SELECT l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity::text, l_extendedprice::text, l_discount::text, l_tax::text, l_returnflag, l_linestatus, l_shipdate, l_commitdate, l_receiptdate, l_shipinstruct, l_shipmode, l_comment FROM tpch10gb.lineitem) TO 's3://copytos3/test/tpch10gb/parquet/lineitem'
WITH (
  AWS CONNECTION = aws_conn,
  MAX FILE SIZE = "100MB",
  FORMAT = 'parquet'
);
rows didn't match; sleeping to see if dataflow catches up
^^^ +++
tpch10gb.td:71:1: error: executing query failed: db error: ERROR: RecordBatch exceeds max file size limit: ERROR: RecordBatch exceeds max file size limit
```
The CSV copy worked fine. Happened in https://github.com/MaterializeInc/materialize/pull/26688 Is it expected and does this mean I should just bump the MAX FILE SIZE parameter? If so, the error message could probably be improved.

rjobanp commented:
> Is it expected and does this mean I should just bump the MAX FILE SIZE parameter? 

@def- nope, you've discovered a bug! It appears that if the max-file-size is below our default record-batch size we will always hit this when the data is large enough. Will fix :)

rjobanp commented:
Just pushed two new commits:

The first implements many of @guswynn suggested comment changes and also fixes the parquet timestamp types we were using to make them more broadly compatible with the parquet ecosystem. This is because I learned that our timestamps only support microsecond-precision, despite being represented by chrono timestamps that support nanosecond precision.

The second changes the logic around dealing with the user-provided max-file-size to be a best-effort approach. This was based on an internal discussion to simplify the uploader logic. I've added to the large docstring for the `ParquetUploader` to describe how it works (or is supposed to work). And I've introduced some `dyncfg` knobs for us to tune the file-size to buffer-size configurations.

guswynn commented on src/aws-util/src/s3_uploader.rs:
```diff
@@ -300,6 +300,7 @@ impl S3MultiPartUploader {
             return Err(S3MultiPartUploadError::ExceedsMaxPartNumber);
         }
 
+        // TODO: Spawn this on a new task to make it parallel.
```
parallel? or just not doing io on timely workers?

rjobanp replied:
it's already not on a timely worker, but we can spawn this in a new task to allow the existing task to continue doing work while this part uploads in the background. There is no need we need to wait for this part to upload here

guswynn commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -455,6 +460,21 @@ where
     completion_stream
 }
 
+/// dyncfg parameters for the copy_to operator, stored in this struct to avoid
+/// requiring storage to depend on the compute crate. See `src/compute-types/src/dyncfgs.rs`
+/// for the definition of these parameters.
+#[derive(Clone, Debug)]
+pub struct CopyToParameters {
```
quite unfortunate but reasonable for now!

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,367 @@
+/// - The [`ParquetFile`] holds a [`ArrowWriter`] that buffers until it has enough data to write
+///   a parquet 'row group'. The 'row group' size is usually based on the number of rows (in the
+///   ArrowWriter), but we also force it to flush based on data-size once it reaches the
+///   [`PARQUET_ROW_GROUP_SIZE`].
```
this doesnt exist anymore right? usually that means the doc build will fail?

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,367 @@
+/// After each row-group is flushed, the size of the file is checked and if it's exceeded
+/// max-file-size a new file is started.
+///
+/// We also set the max ArrowBuilder buffer size to a ratio (e.g. 150%) of the row-group size
+/// to avoid the ArrowWriter buffering too much data itself before flushing a row-group. We're
+/// aiming for the encoded & compressed size of the ArrowBuilder data to be roughly equal to
+/// the row-group size, but this is only an approximation.
```
what happens if we go over this 150% limit?

rjobanp replied:
After the ArrowBuilder gets converted to a RecordBatch and this `RecordBatch` is encoded to a size much larger than the max row-group size, it'll just result in us creating an extra large row-group in the output file. If this happens as the last row-group to be added to the file (since the file goes over the file-size limit), then this could result in us exceeding the max file size by a non-small amount.

I do expect we'll have to tune this ratio to make it work well. I'm going to add an issue to our epic to introduce some test cases that help us tune this parameter by validating how much we end up exceeding the file size limits on large data dumps.

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,367 @@
+            return Ok(());
+        }
+
+        if self.active_file.is_none() {
+            self.start_new_file().await?;
+        }
+
+        if let Some(mut active_file) = self.active_file.take() {
```
if the file is empty, we shouldn't flush it, right?

guswynn commented on src/storage-operators/src/s3_oneshot_sink/parquet.rs:
```diff
@@ -7,28 +7,367 @@
     async fn flush(&mut self) -> Result<(), anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+        self.flush_builder().await?;
+        if let Some(active) = self.active_file.take() {
+            active.finish().await?;
+        }
```
is this not handled in `flush_builder()`?

rjobanp replied:
No it isn't -- this `flush()` method is called when we want to complete the upload, so we need to `finish()` the active file if there is one. The `flush_builder()` method is just for flushing the `ArrowBuilder` to a file.

I think I will rename this method `finish()` as well, since we're essentially 'finishing' the upload completely

rjobanp replied:
actually this made me notice that we weren't running this `finish` call in a tokio task, will fix that too!

guswynn reviewed (approved):
Let me make sure I understand this:

- We add rows to a `ArrowBuilder`.
- When its `row_size > .2 * 1.5 * max_file_size`, we flush it to the `active_file`.
- When the `active_file.in_progress_size()` is `> .2 * max_file_size`, we flush the buffered rows to a row chunk (a `Vec<u8>`) 
- We take that chunk and write it to s3
- We can end up with a file slightly larger than a max file size, but we will make these decisions deterministically across replicas. And now the logic is more clear here, instead of handling errors being returned!

To me this points to the fact that the way the fact that `arrow` and `parquet` crates split `ArrowWriter` and `ArrowBuilder` into pieces and required to the former to take BATCHES of rows instead of just rows is just....not a good API.

I think we can live with this for now, but there are a couple points:
- We should document where we can end up with a slightly too large file (when we do the check against `size_estimate`
- I think splitting `ParquetFile` and `ArrowBuilder` in the `ParquetUploader` is deeply confusing; To me, `ParquetFile` should own the arrow builder, and do all the logic of deciding when to flush to a row batch (at 20% of the file size), when to flush to s3 (at 20% * 150% of the file size), and when to start new files (when we go over the max file size). This change should be a followup (I want this pr in!) but I think it would make everything more readable.

rjobanp commented:
Yes your understanding is correct!

> To me, ParquetFile should own the arrow builder, and do all the logic of deciding when to flush to a row batch (at 20% of the file size), when to flush to s3 (at 20% * 150% of the file size), and when to start new files (when we go over the max file size). This change should be a followup (I want this pr in!) but I think it would make everything more readable.

This is a fantastic idea and should make things simpler. Will address in a followup
